Add support for NSTextInputMarkedRangeAttribute for accessibility clients on the Mac

jhwiig opened the pull request:
#### 94b6fbfe0984cb20f96b9eb6eeac0b6c2d75f1b7
<pre>
Add support for NSTextInputMarkedRangeAttribute for accessibility clients on the Mac
<a href="https://bugs.webkit.org/show_bug.cgi?id=257566">https://bugs.webkit.org/show_bug.cgi?id=257566</a>
rdar://problem/110080851

Reviewed by Tyler Wilcock.

Adding support for NSTextInputMarkedRangeAttribute gives macOS accessibility clients knowledge of where input method events are occurring.
Without this information, it&apos;s impossible for an assistive technology to know if the user is in an input-method composition context.

Test: accessibility/mac/text-input-marked-range.html

* LayoutTests/accessibility/mac/text-input-marked-range-expected.txt: Added.
* LayoutTests/accessibility/mac/text-input-marked-range.html: Added.
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onTextCompositionChange):
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::textInputMarkedRange const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityObjectInterface.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper ALLOW_DEPRECATED_IMPLEMENTATIONS_END]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::setComposition):
* Tools/DumpRenderTree/AccessibilityUIElement.cpp:
(getTextInputMarkedRangeCallback):
(AccessibilityUIElement::getJSClass):
* Tools/DumpRenderTree/AccessibilityUIElement.h:
* Tools/DumpRenderTree/ios/AccessibilityUIElementIOS.mm:
(AccessibilityUIElement::textInputMarkedRange const):
* Tools/DumpRenderTree/mac/AccessibilityUIElementMac.mm:
(AccessibilityUIElement::textInputMarkedRange const):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityUIElementIOS.mm:
(WTR::AccessibilityUIElement::textInputMarkedRange const):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::AccessibilityUIElement::textInputMarkedRange const):

Canonical link: <a href="https://commits.webkit.org/264911@main">https://commits.webkit.org/264911@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/44b87b3caae31c454e5afee4e94ca14757fddc82

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9073 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9360 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9579 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10726 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9029 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9081 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11349 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9328 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11878 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9219 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10190 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7926 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10885 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7485 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8286 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15770 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8594 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8435 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11760 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8940 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7295 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8180 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/8193 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2206 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12399 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8693 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/mac/text-input-marked-range.html:
```diff
@@ -0,0 +1,24 @@
+            shouldBe("element.textInputMarkedRange", "'{0, 5}'");
+        }
+    </script>
+    <script src="../../resources/js-test-post.js"></script>
```
You can remove this line after replacing `js-test-pre.js` with `js-test.js`.

twilco commented on LayoutTests/accessibility/mac/text-input-marked-range.html:
```diff
@@ -0,0 +1,24 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <script src="../../resources/js-test-pre.js"></script>
```
You can use `js-test.js` instead of `js-test-pre.js` + `js-test-post.js`.

twilco commented on LayoutTests/accessibility/mac/text-input-marked-range.html:
```diff
@@ -0,0 +1,24 @@
+        description("This test ensures that the input method marked range is available to accessibility clients.");
+        editable.focus();
+        if (window.textInputController && window.accessibilityController) {
+            var element = accessibilityController.focusedElement;
```
Is setting focus critical to the behavior being tested? If not, `var element = accessibilityController.accessibleElementById("editable")` would be better to use here. Setting focus is async in ITM, so you'd have to do a bit more work to wait for `#editable` to actually become the AX focused element.

jhwiig replied:
I think it's only necessary for setting the marked text. For AX, we can query by ID and that should work all the same. I'll update this. Thanks for making me aware of that ITM case.

aprotyas commented on LayoutTests/accessibility/mac/text-input-marked-range.html:
```diff
@@ -0,0 +1,24 @@
+    </script>
+    <script src="../../resources/js-test-post.js"></script>
+</body>
+</html>
```
Optional nit: trailing newline?

jhwiig replied:
My editor usually does that for me, good catch! (I'll add one)

aprotyas commented on LayoutTests/accessibility/mac/text-input-marked-range.html:
```diff
@@ -0,0 +1,24 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <script src="../../resources/js-test-pre.js"></script>
```
You can just import `<script src="../../resources/js-test.js"></script>` now, no need to import pre/post separately.

jhwiig replied:
Thanks! I didn't know that.

aprotyas commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -404,6 +404,24 @@ std::optional<SimpleRange> AccessibilityObject::misspellingRange(const SimpleRan
+        return std::nullopt;
+
+    auto scope = makeRangeSelectingNodeContents(*node);
+    return characterRange(scope, *range);
```
nit: You can get away with `return { { scope, *range } };`

jhwiig replied:
I'll update it to reflect that, thanks

jhwiig replied:
Unfortunately, I think that this can't be written with list initialization syntax, since technically this function is not a constructor. It's declared at the top of `TextIterator.h` and is designed for converting between `SimpleRange` and `CharacterRange` types. 

If there's something I'm missing that would make this possible though, please let me know! I couldn't get it to compile with this suggested change.

aprotyas replied:
My bad! I assumed you were calling a constructor. This is fine as-is then

jhwiig replied:
All good! There are still many things I have to learn about C++ so these suggestions are always helpful even if they don't pan out.

aprotyas commented on Source/WebCore/accessibility/AccessibilityObject.h:
```diff
@@ -240,6 +240,7 @@ class AccessibilityObject : public AXCoreObject, public CanMakeWeakPtr<Accessibi
     bool hasSameStyle(const AXCoreObject&) const override { return false; }
     bool hasUnderline() const override { return false; }
     bool hasHighlighting() const override;
+    std::optional<CharacterRange> textInputMarkedRange() const override;
```
Can/should this be annotated with `final`?

jhwiig replied:
Yeah, @twilco said that we should be doing that for these types.

aprotyas commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
```diff
@@ -456,6 +456,7 @@ class AXIsolatedObject final : public AXCoreObject {
     bool hasSameStyle(const AXCoreObject&) const override;
     bool hasUnderline() const override { return boolAttributeValue(AXPropertyName::HasUnderline); }
     bool hasHighlighting() const override { return boolAttributeValue(AXPropertyName::HasHighlighting); }
+    std::optional<CharacterRange> textInputMarkedRange() const override { return optionalAttributeValue<CharacterRange>(AXPropertyName::TextInputMarkedRange); }
```
Same question about `final`

aprotyas reviewed (commented):
Some minor, mostly nit comments -- not an exhaustive review!

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -576,6 +576,12 @@ void AXIsolatedTree::updateNodeProperties(AXCoreObject& axObject, const Vector<A
+            if (auto textInputMarkedRange = axObject.textInputMarkedRange())
+                propertyMap.set(AXPropertyName::TextInputMarkedRange, *textInputMarkedRange);
+            else
+                propertyMap.remove(AXPropertyName::TextInputMarkedRange);
```
It seems your intention here is if there is marked text, set the property, and otherwise remove it from the object's property map.

But unfortunately that won't work as you have it now. When you call `propertyMap.remove(AXPropertyName::TextInputMarkedRange)`, you're removing it from the local property map created in the beginning of this method, which is a no-op, and won't result in the property being removed from the object's map when `AXIsolatedTree::m_pendingPropertyChanges` is processed.

What we really need is a mechanism to remove a property from the map as a result of `AXIsolatedTree::updateNodeProperty`. But I don't want to block your change on that, so this should work in the meantime:

```
case AXPropertyName::TextInputMarkedRange: {
    auto textInputMarkedRange = axObject.textInputMarkedRange();
    // FIXME: We should have a mechanism to remove a property from the map for when textInputMarkedRange is std::nullopt.
    propertyMap.set(AXPropertyName::TextInputMarkedRange, textInputMarkedRange ? WTFMove(*textInputMarkedRange) : nullptr);
    break;
}
```

jhwiig replied:
Oops, should have recognized that removing a key from a local here would have no effect when applying it to the original. Thanks for catching that, and I appreciate the suggestion. I'll apply it here.

twilco commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2279,6 +2282,9 @@ void Editor::setComposition(const String& text, const Vector<CompositionUnderlin
         }
     }
 
+    if (auto *cache = m_document.existingAXObjectCache())
```
WebKit C++ coding style requires `auto* foo` instead of `auto *foo`.

jhwiig replied:
Applied to both instances of this. Surprised the linter missed that, it's a habit of my other codebases. Thanks for finding it though.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1580,6 +1580,16 @@ void AXObjectCache::onValidityChange(Element& element)
+{
+#if ENABLE(ACCESSIBILITY_ISOLATED_TREE)
+    if (auto* object = textCompositionObjectForNode(node))
+        updateIsolatedTree(object, AXTextCompositionChanged);
```
Don't need to null-check `object` here (`AXObjectCache::updateIsolatedTree` does it for you). Can do this one-liner instead:

```
updateIsolatedTree(textCompositionObjectForNode(node), AXTextCompositionChanged);
```

jhwiig replied:
Awesome! I am paranoid because of crashes but forgot that if something takes a pointer, it should assume responsibility for checking it.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -4440,7 +4453,22 @@ AXTextChange AXObjectCache::textChangeForEditType(AXTextEditType type)
+
+AccessibilityObject* AXObjectCache::textCompositionObjectForNode(Node& node)
+{
+    // Find the ax object that sends text change notifications for the composition
```
This comment seems redundant thanks to your descriptive function name — can probably remove it.

jhwiig replied:
removed

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -4440,7 +4453,22 @@ AXTextChange AXObjectCache::textChangeForEditType(AXTextEditType type)
+        object = object->observableObject();
+
+#if PLATFORM(MAC)
+    // Fallback on the webarea
```
This comment seems redundant as-is. Maybe it would be good to explain why we only do this for macOS though? If I recall correctly, it's to match behavior elsewhere, so maybe we can call that out.

jhwiig replied:
added some explanation of how we fall back on the Mac for text notifications. It happens in a couple of methods, so I didn't call out any specific one.

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -404,6 +404,39 @@ std::optional<SimpleRange> AccessibilityObject::misspellingRange(const SimpleRan
+    auto* compositionNode = editor.compositionNode();
+    auto* cache = axObjectCache();
+    if (!compositionNode || !cache)
+        return false;
```
To avoid unnecessary work, it would be best to separate these out like so. Also, let's store both in `WeakPtr`s since  we do non-trivial work in this function that could cause them to be destroyed.

```
WeakPtr compositionNode = editor.compositionNode();
if (!compositionNode)
    return false;
WeakPtr cache = axObjectCache();
if (!cache)
    return false;
```

jhwiig replied:
Sure! I have had people ask me to condense these guard statements, but I have no problem expanding them. I also have no idea how much of a cost the cache method call is.

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -404,6 +404,39 @@ std::optional<SimpleRange> AccessibilityObject::misspellingRange(const SimpleRan
+    if (!compositionObject)
+        return false;
+
+    return compositionObject->objectID() == objectID();
```
We could write this a little more simply like so:

```
if (auto* compositionObject = cache->textCompositionObjectForNode(*compositionNode))
    return compositionObject->objectID() == objectID();
return false;
```

jhwiig replied:
suggestion applied!

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -404,6 +404,39 @@ std::optional<SimpleRange> AccessibilityObject::misspellingRange(const SimpleRan
+
+std::optional<CharacterRange> AccessibilityObject::textInputMarkedRange() const
+{
+    auto* node = this->node();
```
This should be a `WeakPtr node` too (sorry for missing this in the last round).

jhwiig replied:
cool, sounds good!

twilco commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2279,6 +2285,16 @@ void Editor::setComposition(const String& text, const Vector<CompositionUnderlin
+            cache->onTextCompositionChange(*previousCompositionNode);
+        // Then update the new cache with current information.
+        if (currentCompositionNode)
+            cache->onTextCompositionChange(*currentCompositionNode);
```
"Clear the previous cache" and "update the new cache" read to me like we're modifying the AXObjectCache, but really we're modifying the compositions for the previous and current nodes. IMO, you could probably just remove these comments — I think your descriptive variable names and function name describe what is happening well enough.

jhwiig replied:
Yeah, just gonna remove them. would like to keep my Editor.cpp footprint low.

jhwiig committed:
Add support for NSTextInputMarkedRangeAttribute for accessibility clients on the Mac
https://bugs.webkit.org/show_bug.cgi?id=257566
rdar://problem/110080851

Reviewed by Tyler Wilcock.

Adding support for NSTextInputMarkedRangeAttribute gives macOS accessibility clients knowledge of where input method events are occurring.
Without this information, it's impossible for an assistive technology to know if the user is in an input-method composition context.

Test: accessibility/mac/text-input-marked-range.html

* LayoutTests/accessibility/mac/text-input-marked-range-expected.txt: Added.
* LayoutTests/accessibility/mac/text-input-marked-range.html: Added.
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onTextCompositionChange):
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::textInputMarkedRange const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityObjectInterface.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper ALLOW_DEPRECATED_IMPLEMENTATIONS_END]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::setComposition):
* Tools/DumpRenderTree/AccessibilityUIElement.cpp:
(getTextInputMarkedRangeCallback):
(AccessibilityUIElement::getJSClass):
* Tools/DumpRenderTree/AccessibilityUIElement.h:
* Tools/DumpRenderTree/ios/AccessibilityUIElementIOS.mm:
(AccessibilityUIElement::textInputMarkedRange const):
* Tools/DumpRenderTree/mac/AccessibilityUIElementMac.mm:
(AccessibilityUIElement::textInputMarkedRange const):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityUIElementIOS.mm:
(WTR::AccessibilityUIElement::textInputMarkedRange const):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::AccessibilityUIElement::textInputMarkedRange const):

Canonical link: https://commits.webkit.org/264911@main
